### PR TITLE
utf16 to utf8 conversion validation

### PR DESCRIPTION
**Pull request by dylan-conway**
Fixes tests with unmatched surrogates. There are still failing tests with `base64`, `base64url`, and error throwing

### CODE REVIEW

File: src/string_immutable.zig [hunk 1]
```diff
@@ -1227,12 +1227,38 @@ pub fn utf16Codepoint(comptime Type: type, input: Type) UTF16Replacement {
     }
 }
 
+pub fn convertUTF16ToUTF8(list_: std.ArrayList(u8), comptime Type: type, utf16: Type) !std.ArrayList(u8) {
+    var list = list_;
+    const length = bun.simdutf.length.utf8.from.utf16.le(utf16);
```
Comment (Jarred-Sumner) [inline src/string_immutable.zig]: This gets the length twice

It should only do it once.

File: src/string_immutable.zig [hunk 2]
```diff
@@ -1227,12 +1227,38 @@ pub fn utf16Codepoint(comptime Type: type, input: Type) UTF16Replacement {
...
+    var start: usize = 0;
+
+    const replacement_char = [_]u8{ 239, 191, 189 };
+    var result = bun.simdutf.convert.utf16.to.utf8.with_errors.le(remaining_input, list.items.ptr[start..list.capacity]);
+    list.items.len = result.count;
+    while (result.status == .surrogate) {
+        try list.resize(list.items.len + 3);
+        start += result.count;
+
+        std.mem.copy(u8, list.items[start..], &replacement_char);
```
Comment (Jarred-Sumner) [inline src/string_immutable.zig]:
```suggestion
      list.items[start..][0..replacement_char.len].* = replacement_char;
```

File: src/string_immutable.zig [hunk 3]
```diff
@@ -1227,12 +1227,38 @@ pub fn utf16Codepoint(comptime Type: type, input: Type) UTF16Replacement {
...
+    try list.ensureTotalCapacityPrecise(length);
+
+    var remaining_input = utf16;
+    var start: usize = 0;
+
+    const replacement_char = [_]u8{ 239, 191, 189 };
+    var result = bun.simdutf.convert.utf16.to.utf8.with_errors.le(remaining_input, list.items.ptr[start..list.capacity]);
+    list.items.len = result.count;
+    while (result.status == .surrogate) {
+        try list.resize(list.items.len + 3);
```
Comment (Jarred-Sumner) [inline src/string_immutable.zig]: This reallocates the array to that specific size

Instead, we want `ensureUnusedCapacity(3)` and then to set `items.len += 3`

File: src/string_immutable.zig [hunk 4]
```diff
@@ -3457,15 +3482,15 @@ pub fn firstNonASCII16CheckMin(comptime Slice: type, slice: Slice, comptime chec
     }
 
     if (comptime check_min) {
-        for (remaining) |char| {
+        for (remaining) |char, i| {
             if (char > 127 or char < 0x20) {
-                return @truncate(u32, (@ptrToInt(std.mem.sliceAsBytes(remaining).ptr) - @ptrToInt(std.mem.sliceAsBytes(slice).ptr)) / 2);
```
Comment (Jarred-Sumner) [inline src/string_immutable.zig:3462]: Was there a bug here or was the code confusing? Adding `i` to this code makes the codegen worse

File: src/string_immutable.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/string_immutable.zig]: I suggest adding a version of this function which doesn't check the length and then having a default version call this function & pass the length in

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
couple comments, mostly worried about the perf impact of checking length twice

File: src/string_immutable.zig (hunk 4 above)
Comment (dylan-conway) [reply src/string_immutable.zig:3462]: There's a bug, this won't return the correct index of the non-ascii character. How bad is codegen with `i`? I could add a counter variable instead

**Comment by Jarred-Sumner** [comment]
Thanks

**Pull request merged**